src/H5Dmpio.c

```diff
@@ -3135,96 +3172,213 @@ H5D__mpio_collective_filtered_chunk_io_setup(const H5D_io_info_t *io_info, const
  *              redistribute (by assigning ownership to a single rank) any
  *              chunks which are selected by more than one MPI rank.
  *
- *              This implementation has MPI rank 0 do all of the chunk
- *              redistribution, which should cut down on MPI communication
- *              overhead, but can potentially cause heavy memory usage on
- *              rank 0. To this end, this strategy should be employed when
- *              the number of chunks being written to is relatively small.
+ *              An initial Allgather is performed to determine how many
+ *              chunks each rank has selected in the write operation and
+ *              then that number is compared against a threshold value to
+ *              determine whether chunk redistribution should be done on
+ *              MPI rank 0 only, or on all MPI ranks.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5D__mpio_redistribute_shared_chunks(H5D_filtered_collective_io_info_t *chunk_list,
+                                     size_t chunk_list_num_entries, const H5D_io_info_t *io_info,
+                                     const H5D_chunk_map_t *fm, int mpi_rank, int mpi_size,
+                                     size_t **rank_chunks_assigned_map)
+{
+    size_t *num_chunks_map = NULL;
+    size_t  coll_chunk_list_size = 0;
+    size_t  i;
+    int     mpi_code;
+    herr_t  ret_value = SUCCEED;
+
+    FUNC_ENTER_STATIC
+
+    HDassert(chunk_list || 0 == chunk_list_num_entries);
+    HDassert(io_info);
+    HDassert(fm);
+    HDassert(mpi_size > 1); /* No chunk sharing is possible for MPI Comm size of 1 */
+    HDassert(rank_chunks_assigned_map);
+
+#ifdef H5Dmpio_DEBUG
+    H5D_MPIO_TRACE_ENTER(mpi_rank);
+    H5D_MPIO_TIME_START(mpi_rank, "Redistribute shared chunks");
+#endif
+
+    /*
+     * Allocate an array for each rank to keep track of the number of
+     * chunks assigned to any other rank in order to cut down on future
+     * MPI communication.
+     */
+    if (NULL == (num_chunks_map = H5MM_malloc((size_t)mpi_size * sizeof(*num_chunks_map))))
+        HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL, "couldn't allocate assigned chunks array")
+
+    /* Perform initial Allgather to determine the collective chunk list size */
+    if (MPI_SUCCESS != (mpi_code = MPI_Allgather(&chunk_list_num_entries, 1, H5_SIZE_T_AS_MPI_TYPE,
```

Comment:
Keep in mind that an initial (all)gather is necessary anyway because someone or everyone needs to know how big of a collective array to allocate for redistribution, so it's not really adding anything in terms of overhead. The allgather here also creates the rank -> num chunks assigned map which saves us two allgather operations later on in the reallocate and reinsert operations. Those used to each do 2 operations - first an allgather to see how many chunks everyone's contributing to the collective array (all of them for linked-chunk I/O, or possibly just 0 or 1 per rank for multi-chunk I/O) and then an allgatherv to actually gather the chunks. The rank -> num chunks assigned map lets us skip the initial allgather in each and just do the allgatherv, so we knock off 2 collective operations.